### PR TITLE
Extract tx confirmation event on `tx raw upgrade-chain` cmd

### PR DESCRIPTION
**Pull request by mzabaluev**
Closes: #1288

## Description

Process the confirmation to the `UpgradeProposal` message as an extended event outside of the `IbcEvent` API.

* Add the `chain::cosmos::events` module to the `relayer` crate.
* In the module, provide a `CosmosEvent` enum for events specific to Cosmos SDK-based chain API.
* Add the `SubmitProposal` variant to `CosmosEvent`, with `SubmitProposal` struct as the value carrying proposal event data.
* Add an inherent method `send_message_and_wait_commit_cosmos` for `CosmosSdkChain`, as a way to wait for an expected Cosmos-specific confirmation event to a message.
* Use this method in the implementation of the `tx raw upgrade-chain` command.
* The internal method `wait_for_block_commits` is modified to try to extract Cosmos-specific events from an ABCI event list, in addition to any IBC events.

______

### PR author checklist:
- [x] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests: integration (for Hermes) or unit/mock tests (for modules). 
- [x] Linked to GitHub issue.
- [x] Updated code comments and documentation (e.g., `docs/`).

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: modules/src/events.rs
Comment (romac) [reply modules/src/events.rs:295]: No idea, let's remove the `Ev` suffix.

File: modules/src/events.rs
Comment (mzabaluev) [inline modules/src/events.rs:346]: I have cargo-culted this from the other event types; not sure if it makes sense in this case.

Comment (adizere) [reply modules/src/events.rs:346]: I have a slight preference to leave `ibc::events::from_tx_response_event` untouched, to remain an IBC-module-specific method, with IBC specific functionality.

Cosmos SDK-specific stuff can live in the `ibc-relayer` crate, since only the relayer needs it (not the IBC module itself). So the suggestion would be to:

- undo the changes in `from_tx_response_event`
- move `modules/src/cosmos/gov/events.rs` -> `relayer/src/chain/cosmos/events.rs`
- modify method [all_ibc_events_from_tx_search_response](https://github.com/informalsystems/ibc-rs/blob/b3f9e4ce39cd3133ae9d88de1541f9f795e9ddc9/relayer/src/chain/cosmos.rs#L2229) to handle two types of events:
    - ibc events, which [it does already](https://github.com/informalsystems/ibc-rs/blob/b3f9e4ce39cd3133ae9d88de1541f9f795e9ddc9/relayer/src/chain/cosmos.rs#L2244-L2247) via method `from_tx_response_event`
    - sdk events, which it can do via the new module `cosmos::events::gov_events::try_from_tx(..)`
- we should also rename 

The solution could look as follows:

```diff
diff --git a/relayer/src/chain/cosmos.rs b/relayer/src/chain/cosmos.rs
index e07f0a5d..f817b167 100644
--- a/relayer/src/chain/cosmos.rs
+++ b/relayer/src/chain/cosmos.rs
@@ -2226,7 +2226,7 @@ fn update_client_from_tx_search_response(
         .map(IbcEvent::UpdateClient)
 }

-fn all_ibc_events_from_tx_search_response(chain_id: &ChainId, response: ResultTx) -> Vec<IbcEvent> {
+fn all_events_from_tx_search_response(chain_id: &ChainId, response: ResultTx) -> Vec<IbcEvent> {
     let height = ICSHeight::new(chain_id.version(), u64::from(response.height));
     let deliver_tx_result = response.tx_result;
     if deliver_tx_result.code.is_err() {
@@ -2245,6 +2245,10 @@ fn all_ibc_events_from_tx_search_response(chain_id: &ChainId, response: ResultTx
             trace!("got IBC event: {:#?}", ibc_ev);
             result.push(ibc_ev);
         }
+        if let Some(sdk_ev) = cosmos::events::sdk_events::from_tx_response_event(height, &event) {
+            trace!("got SDK event: {:#?}", ibc_ev);
+            result.push(sdk_ev);
+        }
     }
```

**Review by adizere — CHANGES_REQUESTED** [review]
Have not tested yet. Left an idea for a possible improvement to make the separation IBC <> SDK cleaner. The idea is to keep the `ibc` crate purely about IBC. It's ok to have SDK-related logic in the `ibc-relayer`, so that's where the new module `gov/events.rs` should land.

**Comment by romac** [comment]
Maybe we should start looking into https://github.com/informalsystems/ibc-rs/issues/1580?

**Comment by adizere** [comment]
> Maybe we should start looking into #1580?

Indeed. That issue may be a big lift. We'll need it for a stable v1 release. I think PRs such as the present one and #1845 are nice sub-problems of #1580, so its a good exercise to fix them before that one.

File: modules/src/events.rs
Comment (mzabaluev) [reply modules/src/events.rs:346]: The difficulties I've run into with the recommended approach is, the result of `ChainHandle::query_txs` et al. is currently `Vec<IbcEvent>`, and remaking all of these uses to return a new combined event type is quite a slog. @adizere do you think we should commit to such a redesign, or could there be another way to solve this?

I've tried to come up with a simpler solution, special-casing the SDK event handling aside from the generic `ChainHandle` API and adding a cosmos-specific method to call directly, but such a refactor looks hackish and I feel that we should not do this for every code path where an SDK-specific event might be used to confirm a transaction.

**Comment by mzabaluev** [comment]
This is turning into a bigger rework, with the following options:

| Solution | Advantages | Problems |
| ----------- | --------------- | ------------- |
| Extend `IbcEvent` with Cosmos-specific variants | Already implemented in the branch | Breaks modularity, adds Cosmos specific data to `ibc` |
| Enum unifying `IbcEvent` with `SdkEvent` in `relayer` code | Modularity preserved | Lots of tedious changes to code that only really needs `IbcEvent` at this point |
| Extend `IbcEvent` with a dynamically encoded variant to pass custom events back to CLI | Modularity preserved | Soft-typed data added to the API |

**Comment by adizere** [comment]
Thank you for the breakdown Mikhail! This is helpful. I started out assuming option 2 is ideal, but it may be more wise to proceed with 1 (or even 3) and commit to a refactor to `IbcEvent` later.

**Comment by romac** [comment]
Agreed, I vote for option 1. Once we get to add proper support for non-Cosmos chains (together with https://github.com/informalsystems/ibc-rs/issues/1580), we can move the SDK-specific event out of the `ibc` crate. Let's just make sure to document that this is temporary and that non-Cosmos chains which implement IBC using the `ibc` crate need not care about this event.

**Comment by romac** [comment]
You mean #1979, right?

**Comment by mzabaluev** [comment]
I have reworked the solution, taking none of the approaches discussed in https://github.com/informalsystems/ibc-rs/pull/1909#issuecomment-1065095205, but instead adding an entirely different path for receiving Cosmos-specific confirmation events.

File: modules/src/events.rs
Comment (adizere) [inline modules/src/events.rs:275]: Intersting. I wonder why are these lines marked as changed since I don't think there should be any diff.

**Review by adizere — COMMENTED** [review]
I went through this quickly and I have some minor remarks only.

Still testing.

> I have reworked the solution, taking none of the approaches discussed in https://github.com/informalsystems/ibc-rs/pull/1909#issuecomment-1065095205, but instead adding an entirely different path for receiving Cosmos-specific confirmation events.

Considering this new approach, I’m wondering what is the best way to choose a winner between #1979 and #1909. Any ideas or recommendations?

**Comment by mzabaluev** [comment]
> Considering this new approach, I’m wondering what is the best way to choose a winner between https://github.com/informalsystems/ibc-rs/pull/1979 and https://github.com/informalsystems/ibc-rs/pull/1909. Any ideas or recommendations?

This one (#1909) is meant to "get the job done" as per the original issue: wait for a confirmation event and print it out.

There are still issues with it, though: as `query_txs` has been left unmodified, the `query tx events` command would not print the extra events.

**Comment by mzabaluev** [comment]
> There are still issues with it, though: as `query_txs` has been left unmodified, the `query tx events` command would not print the extra events.

This can also can be modified to use the `query_extended_events_for_tx`: https://github.com/informalsystems/ibc-rs/blob/1ca72521d3243de7af8d809b87f3c5dfe625cb7c/relayer-cli/src/commands/query/tx/events.rs#L55
But I'm not sure about how to modify the output: a tuple of two event arrays? Or a struct-like output with `ibc_events` and `cosmos_events` fields? Maybe some flattened event array (partially going back to the option 2 from https://github.com/informalsystems/ibc-rs/pull/1909#issuecomment-1065095205)?

File: relayer/src/chain/cosmos.rs
Comment (mzabaluev) [reply relayer/src/chain/cosmos.rs:882]: I have reworked this to only perform a single query and extract both kinds of events from the response.
Updated the comments accordingly.

File: modules/src/events.rs
Comment (mzabaluev) [reply modules/src/events.rs:275]: This removes the unnecessary "Ev" suffixes from the text repr, as discussed in https://github.com/informalsystems/ibc-rs/pull/1909#discussion_r812036743

File: relayer/src/chain/cosmos.rs
Comment (seanchen1991) [inline relayer/src/chain/cosmos.rs:2462]: Now that there's a `cosmos_events` field, this field should also probably be named more explicitly. 
```suggestion
    ibc_events: Vec<IbcEvent>,
```

File: relayer/src/chain/cosmos/events.rs
Comment (seanchen1991) [inline relayer/src/chain/cosmos/events.rs:1]: I'm not quite sure what "followed via Cosmos SDK" means. Would "chains that adhere to the Cosmos SDK" be more accurate?

File: relayer/src/chain/cosmos/events.rs
Comment (seanchen1991) [inline relayer/src/chain/cosmos/events.rs:13]:
```suggestion
/// Variants that encode the possible message types sent via Cosmos SDK modules
#[derive(Debug, Serialize)]
```

File: relayer/src/chain/cosmos/events.rs
Comment (seanchen1991) [inline relayer/src/chain/cosmos/events.rs:15]:
```suggestion
    /// An event sent via the governance module indicating a new governance proposal was submitted
    SubmitProposal(SubmitProposal),
```

**Review by seanchen1991 — CHANGES_REQUESTED** [review]

File: relayer/src/chain/cosmos.rs
Comment (mzabaluev) [reply relayer/src/chain/cosmos.rs:965]: Good point, though I suspect the win will be negligible in most cases.

File: relayer/src/chain/cosmos/events.rs
Comment (romac) [reply relayer/src/chain/cosmos/events.rs:1]: _chains built on the Cosmos SDK_ ?

File: relayer/src/chain/cosmos/events.rs
Comment (romac) [inline relayer/src/chain/cosmos/events.rs:37]: Let's introduce a new type for this?

Comment (mzabaluev) [reply relayer/src/chain/cosmos/events.rs:37]: I'd rather not overdo it at this point, since we don't have any use of the field besides dumping it in the command output.

**Comment by mzabaluev** [comment]
The clippy lints are unrelated, will be fixed by #2079.

File: relayer/src/chain/cosmos.rs
Comment (seanchen1991) [inline relayer/src/chain/cosmos.rs:2456]:
```suggestion
    /// The IBC events generated by a Tx once executed
```

File: relayer/src/chain/cosmos.rs
Comment (seanchen1991) [inline relayer/src/chain/cosmos.rs:2458]:
```suggestion
    /// Cosmos-specific events generated by a Tx once executed
```

File: relayer/src/chain/cosmos.rs
Comment (seanchen1991) [inline relayer/src/chain/cosmos.rs:2454]:
```suggestion
    /// The broadcast_tx_sync response
```

**Review by seanchen1991 — APPROVED** [review]
Just a few nits. Otherwise, looks good to me 🙂

File: relayer/src/chain/cosmos.rs
Comment (adizere) [inline relayer/src/chain/cosmos.rs:975]: Do I understand correctly that the only difference between this method and `send_messages_and_wait_commit` is that the former accepts a single `msg: Any`, whereas the latter accepts as parameter a `msgs: TrackedMsgs` (which it splits across multiple tx-es).

The comment and method name here implies that is some more significant difference to the other `send_messages_and_wait_*` method.

**Review by adizere — COMMENTED** [review]
I left a couple of questions that should be addressed or clarified before merging.

Given the complexity of changes in the present PR, the fix suggested in #1979 is simpler and does a good enough job to accept it as a solution for #1288. We should probably accept the current PR if it gives us insights, if it offers a useful refactoring, or if it adds support that might be needed in the future, eg non-SDK support or helps in a way with integration tests; I don't understand if any of these are true at the moment. That's my current assessment.

I think Romain & Soares are in the best position to choose the right fix between PRs #1979 and #1909, so I will defer to them. (Again, both PRs "solve the problem" from a user-level perspective).

File: relayer/src/chain/cosmos.rs
Comment (mzabaluev) [reply relayer/src/chain/cosmos.rs:873]: >     1. Does this mean that _all_ Hermes CLIs will do `query_extended_events_for_tx` as of this PR?

Yes. I chose this as a way to avoid copying `wait_for_block_commits` to do almost exactly the same job, while not exposing these internal changes in the generic `send_messages_and_wait_commit` API.

>     2. Does this have any performance implications or does it affect the return value for any other Hermes CLI command (aside from `tx raw upgrade chain)?

I expect the performance implications to be negligible: Cosmos and IBC events are [parsed out of the same list](https://github.com/informalsystems/ibc-rs/blob/fd3d09e717df8f7d480b11baba25fe73d9974318/relayer/src/chain/cosmos.rs#L965) of ABCI events, which is still retrieved in a single query (in an intermediate change it was not so, but I have reworked it to avoid this exact performance hit).

As for the return value, the only way it would change the output of any other CLI command or other relayer logic using `send_messages_and_wait_commit`, is if the caller expected an IBC event in response to the message, but got no IBC events and some recognized Cosmos event(s) instead. In this case, the return value would be an empty list (as the Cosmos events get discarded in the generic API). I expect there are no such cases at the moment, and if any appear, that would indicate something wrong in our expectations about the ABCI chain.

File: relayer/src/chain/cosmos.rs
Comment (mzabaluev) [reply relayer/src/chain/cosmos.rs:975]: I did not want to replicate the message batching code, since the only current use of this method sends a single message.
Maybe this needs to be clarified in the documentation.

**Comment by adizere** [comment]
Closing this in favor of the simpler solution offered in #1979.

We will revive this code as the need arises in the future. In particular:

- the method `query_extended_events_for_tx` would come in handy when we will require waiting on non-IBC events;
- the method `send_message_and_wait_commit_cosmos` would come in handy when we will require a separate path for submitting transactions that is SDK-specific, which short-circuits the `ChainHandle` interface and could also be useful in testing frameworks.

**Pull request closed without merging**
